pkg/probe/probes.go

```diff
@@ -96,54 +118,144 @@ func checkNodeReadiness(client client.Client, timeout time.Duration) error {
 			Get("routes.running.#(destination==\"0.0.0.0/0\").next-hop-address").String()
 		if defaultGw == "" {
-			log.Info("default gw missing", "state", string(currentState))
+			log.Info("default gw missing", "state", gjsonCurrentState.String())
 			return false, nil
 		}
 
 		return true, nil
 	})
 }
 
-func RunAll(client client.Client) error {
+func runPing(client client.Client, timeout time.Duration) error {
 	defaultGw, err := defaultGw()
 	if err != nil {
 		return errors.Wrap(err, "failed to retrieve default gw at runProbes")
 	}
 
-	currentState, err := nmstatectl.Show()
+	pingOutput, err := ping(defaultGw, timeout)
 	if err != nil {
-		return errors.Wrap(err, "failed to retrieve currentState at runProbes")
+		return errors.Wrapf(err, "error pinging default gateway -> output: %s", pingOutput)
 	}
-
-	// TODO: Make ping timeout configurable with a config map
-	pingOutput, err := ping(defaultGw, defaultGwProbeTimeout)
+	return nil
+}
+func lookupRootNS(nameServer string, timeout time.Duration) error {
+	rootNS := "root-server.net"
+	r := &net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			d := net.Dialer{
+				Timeout: timeout,
+			}
+			return d.DialContext(ctx, network, net.JoinHostPort(nameServer, "53"))
+		},
+	}
+	// We use a closure to create a scope for defer here
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	_, err := r.LookupNS(ctx, rootNS)
 	if err != nil {
-		return errors.Wrapf(err, "error pinging external address after network reconfiguration -> output: %s, currentState: %s", pingOutput, currentState)
+		return errors.Wrapf(err, "failed looking up NS %s using name sever %s", rootNS, nameServer)
 	}
+	return nil
+}
 
-	err = checkApiServerConnectivity(apiServerProbeTimeout)
+func runDNS(client client.Client, timeout time.Duration) error {
+	currentStateAsGJson, err := currentStateAsGJson()
 	if err != nil {
-		return errors.Wrapf(err, "error checking api server connectivity after network reconfiguration -> currentState: %s", currentState)
+		return errors.Wrap(err, "failed retrieving current state to get name resolving config")
 	}
 
-	err = checkNodeReadiness(client, nodeReadinessProbeTimeout)
+	// Get the name servers at node since the ones at container are not accurate
+	runningServersGJsonPath := "dns-resolver.running.server"
+	runningNameServers := currentStateAsGJson.Get(runningServersGJsonPath).Array()
+	if len(runningNameServers) == 0 {
+		return fmt.Errorf("missing name servers at '%s' on %s", runningServersGJsonPath, currentStateAsGJson.String())
+	}
+
+	// Check that at least one of the name servers can resolve root server NS
+	resolved := false
+	for _, runningNameServer := range runningNameServers {
+		err = lookupRootNS(runningNameServer.String(), defaultDnsProbeTimeout)
+		if err != nil {
+			log.Info("WARNING cannot resolve root server NS", "err", err)
+		} else {
+			resolved = true
```

Comment:
You can return nil here and get rid of the resolved variable. It's a matter of taste though. The current state is fine too.

Reply:
I want this to fail if there is no resolve names so DNS probe is not selected if there is none before apply config, for that I need the `resolved` variable that will be false if code does not enter for loop.